src/include/storage/index/hash_index.h

```diff
@@ -0,0 +1,272 @@
+        return false;
+      };
+
+      const bool UNUSED_ATTRIBUTE uprase_result = hash_map_->uprase_fn(index_key, abort_key_found_fn);
+      TERRIER_ASSERT(!uprase_result, "This operation should NOT insert a new key into the cuckoohash_map.");
+    });
+
+    return true;
+  }
+
+  bool InsertUnique(transaction::TransactionContext *const txn, const ProjectedRow &tuple,
+                    const TupleSlot location) final {
+    TERRIER_ASSERT(GetConstraintType() == ConstraintType::UNIQUE,
+                   "This Insert is designed for indexes with uniqueness constraints.");
+    KeyType index_key;
+    index_key.SetFromProjectedRow(tuple, metadata_);
+    bool predicate_satisfied = false;
+
+    // The predicate checks if any matching keys have write-write conflicts or are still visible to the calling txn.
+    auto predicate = [txn](const TupleSlot slot) -> bool {
+      const auto *const data_table = slot.GetBlock()->data_table_;
+      const auto has_conflict = data_table->HasConflict(*txn, slot);
+      const auto is_visible = data_table->IsVisible(*txn, slot);
+      return has_conflict || is_visible;
+    };
+
+    bool UNUSED_ATTRIBUTE insert_result = false;
+
+    auto key_found_fn = [location, &insert_result, &predicate_satisfied, predicate](ValueType &value) -> bool {
+      if (std::holds_alternative<TupleSlot>(value)) {
+        const auto existing_location = std::get<TupleSlot>(value);
+        predicate_satisfied = predicate(existing_location);
+        if (!predicate_satisfied) {
+          // existing location is not visible, replace with a ValueMap
+          value = ValueMap({{location}, {existing_location}}, 2);
+          insert_result = true;
+        }
+      } else {
+        auto &value_map = std::get<ValueMap>(value);
+
+        for (const auto i : value_map) {
+          predicate_satisfied = predicate_satisfied || predicate(i);
+        }
+
+        if (!predicate_satisfied) {
+          // insert the location to the cuckoohash_map
+          insert_result = value_map.emplace(location).second;
+          TERRIER_ASSERT(insert_result,
+                         " index shouldn't fail to insert after predicate check. If it did, something went wrong deep "
+                         "inside the hash map itself.");
+        }
+      }
+      return false;
+    };
+
+    const bool UNUSED_ATTRIBUTE uprase_result = hash_map_->uprase_fn(index_key, key_found_fn, location);
+
+    TERRIER_ASSERT(predicate_satisfied || (insert_result != uprase_result),
+                   "Either a new key was inserted (uprase_result), or the value already existed and a new value was "
+                   "inserted (insert_result).");
+
+    const bool UNUSED_ATTRIBUTE overall_result = insert_result || uprase_result;
+
+    if (overall_result) {
+      txn->RegisterAbortAction([=]() {
+        auto abort_key_found_fn = [location](ValueType &value) -> bool {
```

Comment:
Maybe this lambda should be turned into a macro within this file.  It's the same logic that has to be inlined so I think only a macro would let you write it once.

Reply:
Yeah I'm considering how macro-heavy I want to go with this stuff. If I define them all in index.h I can use them in both derived index wrappers, but using macros feels so C99.